drafts/records.md

```diff
@@ -0,0 +1,398 @@
+$pluto = Planet("Pluto", 0);
+// we made it!
+$pluto = $pluto->with(population: 1);
+// and then we changed the name
+$mickey = $pluto->with(name: "Mickey"); // no named argument for population error
+```
+
+#### Constructors
+
+Optionally, they may also define a constructor to provide validation or other initialization logic:
+
+```php
+record User(string $name, string $email) {
+  public string $id;
+
+  public function __construct() {
+    if (!filter_var($email, FILTER_VALIDATE_EMAIL)) {
+      throw new InvalidArgumentException("Invalid email address");
+    }
+    
+    $this->id = hash('sha256', $email);
+    $this->name = ucwords($name);
+  }
+}
+```
+
+During construction, a `record` is fully mutable.
+This allows the developer freedom to mutate properties as needed to ensure a canonical representation of an object.
+
+### Performance considerations
+
+To ensure that records are both performant and memory-efficient,
+the RFC proposes leveraging PHP's copy-on-write (COW) semantics (similar to arrays) and interning values.
+Unlike interned strings, the garbage collector will be allowed to clean up these interned records when they are no
+longer needed.
+
+``` php
+$point1 = Point(3, 4);
+$point2 = $point1; // No data duplication, $point2 references the same data as $point1
+$point3 = Point(3, 4); // No data duplication here either, it is pointing the the same memory as $point1
+
+$point4 = $point1->with(x: 5); // Data duplication occurs here, creating a new instance with modified data
+```
+
+#### Cloning and with()
+
+Calling `clone` on a `record` results in the same record object being returned. As it is a "value" object, it
+represents a value and is the same thing as saying `clone 3`—you expect to get back a `3`.
+
+`with` may be called with no arguments, and it is the same behavior as `clone`.
+This is an important consideration because a developer may call `$new = $record->with(...$array)` and we don’t want to
+crash.
+If a developer wants to crash, they can do by `assert($new !== $record)`.
+
+### Serialization and deserialization
+
+Records are fully serializable and deserializable.
+
+```php
+record Single(string $value);
+record Multiple(string $value1, string $value2);
+
+echo $single = serialize(Single('value')); // Outputs: "O:6:"Single":1:{s:5:"value";s:5:"value";}"
+echo $multiple = serialize(Multiple('value1', 'value2')); // Outputs: "O:8:"Multiple":1:{s:6:"values";a:2:{i:0;s:6:"value1";i:1;s:6:"value2";}}"
+
+echo unserialize($single) === Single('value'); // Outputs: true
+echo unserialize($multiple) === Multiple('value1', 'value2'); // Outputs: true
+```
+
+### Equality
+
+A `record` is always strongly equal (`===`) to another record with the same value in the properties,
+much like an `array` is strongly equal to another array containing the same elements.
+For all intents, `$recordA === $recordB` is the same as `$recordA == $recordB`.
+
+Comparison operations will behave exactly like they do for classes, for example:
```

Comment:
For those who do not know exactly how classes compare — since I never found the need to compare them — maybe elaborate a bit more here?

Reply:
It's an undocumented "feature" so I'm not even sure what it means myself. I'm just calling out that there isn't a behavior change here.